### DISCUSSION TITLE
ZodObject not assignable to ZodTypeAny

### DISCUSSION THREAD

**Discussion by cbosman-docksters** [colinhacks/zod#1993] [Q&A] [answered]
I've encountered this issue since updating to 3.20.2 (have also tried going down to 3.20.1 and 3.20.0, but it only disappears in 3.19), where I get ```"Argument of type ZodObject<....> is not assignable to parameter of type ZodTypeAny. Type ZodObject<...> is missing the following properties from 'ZodType<any, any, any>': catch, pipe.```

I've tried searching for a solution to the issue but cannot seem to find anything.

Reproduction steps:
- Create a file `1.ts`
- Add some sort of zod object to this file. `export const a = z.object({ ... });`
- Then import said zod object into another file `2.ts`, where you try to use it in another zod object (so basically nesting them).
- `const b = z.object({ someName: a })`

This worked in previous versions but I just cannot seem to get it working in this update. I'd love to hear any feedback on this, and maybe hear if I'm doing anything wrong or if it's a known issue.

Thanks!

**Comment by Tanimodori** [comment]
Version 3.20.0 definitely breaks something with `ZodObject` types. As type inferring in #1955 works in `3.19.1` but not in `^3.20.0`.

**Comment by Tanimodori** [comment]
Seems unrelated to https://github.com/colinhacks/zod/issues/1955. I cannot reproduce this issue here. Could you provide a minimal reproduction please?

**Comment by JacobWeisenburger** [comment]
@cbosman-docksters 
Everything seems to be working fine for me. Please provide a minimal reproduction of your issue.
```ts
const a = z.object( { foo: z.string() } )
const b = z.object( { someName: a } )
console.log( b.parse( { someName: { foo: 'foo' } } ) )
// { someName: { foo: 'foo' } }
```

**Comment by akomm** [comment]
Make sure you have set `strict: true` in `tsconfig.json`'s `compilerOptions`. It is a requirement for zod to work correctly. If the problem persists afterwards, also provide an example for reproduction as suggested above.

**Comment by cbosman-docksters** [comment]
Hi there. I do have `strict: true` in my tsconfig.json.

Reproduction:

```1.ts
export const foo = z.object({
    test: z.string(),
});

2.ts
import {foo} from "1.ts";
const foobar = z.object({
        bar: foo,
});
```

This is the output of the ts compiler:

` Type 'ZodObject<{ test: ZodString; }, "strip", ZodTypeAny, { test: string; }, { test: string; }>' is missing the following properties from type 'ZodType<any, any, any>': catch, pipe`

**Comment by JacobWeisenburger** [comment, marked as answer]
I have tried this, and everything is working as expected.

```ts
// 1.ts
import { z } from 'zod'

export const foo = z.object( {
    test: z.string(),
} )
```

```ts
// 2.ts
import { foo } from "./1.ts"
const foobar = z.object( {
    bar: foo,
} )
// no errors
```

Can you please provide your package.json and tsconfig.json?

Alternatively, you could provide a link to a codesandbox or something that has everything the way you do.

**Comment by cbosman-docksters** [comment]
I use it in a monorepo via yarn workspaces, could that have anything to do with it? I indeed cannot seem to reproduce it in codesandbox.

package.json
```
{
  "name": "internal",
  "version": "1.0.0",
  "description": "",
  "type": "module",
  "scripts": {
    "test": "jest"
  },
  "license": "UNLICENSED",
  "private": true,
  "dependencies": {
    "@aws-sdk/util-dynamodb": "^3.215.0",
    "common": "*",
    "dayjs": "^1.11.7",
    "fp-ts": "^2.13.1",
    "geolib": "^3.3.3",
    "io-ts": "^2.2.19",
    "jimp": "^0.16.2",
    "serverless-offline": "^11.6.0",
    "svgo": "2.8.0",
    "tiny-async-pool": "^2.1.0",
    "tsup": "^6.3.0",
    "zod": "3.20.2"
  },
  "devDependencies": {
    "@aws-sdk/client-cognito-identity-provider": "^3.216.0",
    "@aws-sdk/client-dynamodb": "^3.216.0",
    "@aws-sdk/client-s3": "^3.216.0",
    "@aws-sdk/client-ses": "^3.216.0",
    "@types/aws-lambda": "^8.10.108",
    "@types/jest": "^29.4.0",
    "@types/node": "^18.11.9",
    "@types/svgo": "^2.6.4",
    "@types/tiny-async-pool": "^2.0.0",
    "serverless": "^3.24.1",
    "serverless-plugin-typescript": "^2.1.4",
    "serverless-prune-plugin": "^2.0.1",
    "ts-loader": "^9.4.1",
    "ts-node": "^10.9.1",
    "typescript": "^4.8.4"
  }
}
```

tsconfig.json
```
{
    "compilerOptions": {
        "noImplicitAny": true,
        "removeComments": true,
        "outDir": "dist",
        "sourceMap": false,
        "moduleResolution": "node",
        "module": "esnext",
        "declaration": true,
        "declarationMap": true,
        "strict": true,
        "downlevelIteration": true,
        "resolveJsonModule": true,
        "target": "esnext",
        "paths": {
            "/opt/nodejs/*": ["../packages/*"]
        },
        "lib": ["dom", "dom.iterable", "esnext"],
        "allowSyntheticDefaultImports": true
    },
    "include": ["src/**/*", "index.d.ts"]
}
```

**Comment by cbosman-docksters** [comment]
Nvm, I'm a dumbass. I was mixing-and-matching zod versions in my codebase. In this case, I was using this "common" package I built myself, which was on an older version of zod (namely 3.19) and trying to use those older validators in the newer 3.20 version of zod. Thanks for helping me!

**Reply by realwitt** [reply]
This was also my problem with mixing v3 and v4... much appreciated 🫡

**Comment by JacobWeisenburger** [comment]
If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
